data/tutorials/language/0lg_10_polymorphic_variants.md

```diff
@@ -0,0 +1,713 @@
+```ocaml
+# type color = [ `Gray of int | `RGB of int * int * int ];;
+type color = [ `Gray of int | `RGB of int * int * int ]
+
+# type extended_color = [ color | `RGBA of int * int * int * int ];;
+type extended_color =
+    [ `Gray of int | `RGB of int * int * int | `RGBA of int * int * int * int ]
+
+# let color_to_int = function
+  | `RGB (r, g, b) -> 36 * r + 6 * g + b + 16
+  | `Gray i -> i + 232;;
+val color_to_int : [< `Gray of int | `RGB of int * int * int ] -> int = <fun>
+
+# let extended_color_to_int = function
+  | `RGBA (r, g, b, a) -> 216 * r + 36 * g + 6 * b + a + 16
+  | `Grey i -> i + 2000
+  | #color as color -> color_to_int color;;
+val extended_color_to_int :
+  [< `Gray of int
+   | `Grey of int
+   | `RGB of int * int * int
+   | `RGBA of int * int * int * int ] ->
+  int = <fun>
+
+```
+
+The function `color_to_int` can convert `` `RGB`` or `` `Gray`` values. The function `extended_color_to_int` is intended to convert `` `RGB``, `` `Gray`` or `` `RGBA`` values; but it is supposed to apply a different conversion formula for gray scales. However, a typo was made, it is spelled `` `Gray``. The type-checker accepts this definition of `extended_color_to_int` as a function accepting four tags.
+
+### Over Approximating Type
+
+The following function was presented in the [Inferred Type Aliases](#Inferred-Type-Aliases) section.
+```ocaml
+# function `Avocado -> `Cilantro | plant -> plant;;
+- : ([> `Cilantro | `Avocado ] as 'a) -> 'a = <fun>
+```
+
+Because of the `plant -> plant` pattern, the types inferred as domain and codomain are the same. As `` `Avocado`` is accepted, it must be part of the domain; and as`` `Cilantro`` is returned, it must be part of the codomain. Both end up being part of the common type. However, `` `Avocado`` should not be part of the codomain, as this function can't possibly return this value. A finer type-checker would infer more precise types. Type-checking is an approximation and a trade-off, some valid programs are rejected, and some types are too coarse.
+
+### Performances
+
+TODO: Expand this section
+
+> There is one more downside: the runtime cost. A value `Pair (x, y)` occupies 3 words in memory, while a value `` `Pair (x, y)`` occupies 6 words.
+
+Guillaume Melquiond
+
+## When to Use Polymorphic Variant
+
+Only using polymorphic variants is over-engineering, it should be avoided. Polymorphic variants aren't an extension of simple variants. From the data perspective, polymorphic variants and simple variants are equivalent. They both are sum types in the algebraic data types sense, both with parametric polymorphism and recursion. Back quote isn't a difference that matters. The only meaningful difference is the algorithm. Therefore, deciding when to use polymorphic variants boils down to another question:
```

Comment:
> "Only using polymorphic variants is over-engineering,"
I also don't recommend people to _only use polyvars_ but I think we can do better and explain _why_.

For example, the type errors can get VERY out of hand.

Also not sure why reiterating the whole "not an extension, but equivalent" is relevant.

Reply:
Do you have an example code triggering such type errors? I may be useful. 

"not an extension, but equivalent": Yes, and that discussion will by downsized and what's left will move to the end of the document.